### PR TITLE
Revise query handling for Datastore (again)

### PR DESCRIPTION
**Pull request by jskeet**
This is a second attempt at improving the query experience for Datastore, building on the comments received on #292. I think it's now much simpler to use for web applications - Jeff, I'm happy to create a PR to show how it would change your sample code, although I guess one good test would be to see whether it's discoverable enough for you to do it :)

I'm much happier with this version.

Commit message:

- Rename DatastoreQueryResults to LazyDatastoreQuery
  for clarity
- Add DatastoreQueryResults which is just an in-memory
  set of results, a cursor and an end condition
- Add LazyDatastoreQuery.GetAllResults() to fetch
  everything (ideal for web apps)
- Remove AsEntityResults() as it's rarely useful now
- Remove AsBatches() as it's rarely useful now
- Reword documentation for AsResponses() to indicate
  that it's a pretty niche use case.
- Remove details from DatastoreDb method calls, as
  there's now more documentation in LazyDatastoreQuery
  and the name is more appropriate.
- Ditto everything async

Basically, most use cases will fall into "stream all
results, I don't care about cursors" (just iterate
over LazyDatastoreQuery) or "I'm setting the limit
myself, get all the results" (call GetAllResults()).

### CODE REVIEW

File: apis/Google.Datastore.V1Beta3/Google.Datastore.V1Beta3.Snippets/DatastoreDbSnippets.cs [hunk 1]
```diff
@@ -90,10 +91,11 @@ public void StructuredQuery()
             {
                 Filter = Filter.Equal("author", "Jane Austen")
             };
-            DatastoreQueryResults results = db.RunQuery(query);
-            // DatastoreQueryResults implements IEnumerable<Entity>, but you can
-            // call AsEntityResults(), AsBatches() or AsResponses() to see the query
-            // results in whatever way makes most sense for your application.
+            LazyDatastoreQuery results = db.RunQuery(query);
```
Comment (SurferJeffAtGoogle) [inline apis/Google.Datastore.V1Beta3/Google.Datastore.V1Beta3.Snippets/DatastoreDbSnippets.cs]: The word "Lazy" in LazyDatastoreQuery helps communicate to readers the semantics of the query, but most of the time, the name of the type will not be visible.  9 times out of 10, I will write:
  var results = db.RunQuery(query);

Comment (mteper) [inline apis/Google.Datastore.V1Beta3/Google.Datastore.V1Beta3.Snippets/DatastoreDbSnippets.cs]: Possible alternatives:
- `.RunLazyQuery()`
- `.RunQueryLazy()`
- `.LazyQuery()`

Comment (SurferJeffAtGoogle) [inline apis/Google.Datastore.V1Beta3/Google.Datastore.V1Beta3.Snippets/DatastoreDbSnippets.cs]: I imagine it's quite frustrating to hear me offer alternative after alternative on this point, but I think I finally found one where the users assumptions match the actual RPCs generated.

db.RunQuery(query) returns all the results full fetched, like db.RunQuery.GetAllResults() does currently.
db.IterateQuery(query) returns an IEnumerator<T> (not an IEnumerable<T>).

Then, my assumptions match the underlying implementation.  Specifically:
1. RunQuery() fetches all the results now.
2. IterateQuery() starts fetching results and will fetch more as I iterate over the results.
3. I will never unknowingly initiate an RPC by iterating over results.

Comment (mteper) [inline apis/Google.Datastore.V1Beta3/Google.Datastore.V1Beta3.Snippets/DatastoreDbSnippets.cs]: I think there's a problem with discoverability with these two methods. Users will look for and discover `RunQuery` but won't come across `IterateQuery` next to it in IntelliSense. I would still favor something like RunQuery / RunQueryLazy for proximity.

Comment (jskeet) [inline apis/Google.Datastore.V1Beta3/Google.Datastore.V1Beta3.Snippets/DatastoreDbSnippets.cs]: The reason for keeping RunQuery as the method name was to connect with the RPC API - any users who look at the RPCs (or other platforms exposing the RPCs directly) is likely to look for RunQuery.

I wouldn't mind an API on DatastoreDb of:

```csharp
DatastoreQueryResults RunQueryEagerly(...)
LazyDatastoreQuery RunQueryLazily(...)
```

... and we could change LazyDatastoreQuery to make *that* the point at which you decide sync/async, as well as the view you want:

```csharp
IEnumerable<Entity> GetEntities()
IAsyncEnumerable<Entity> GetEntitiesAsync()
IEnumerable<QueryResponse> GetRpcResponses()
IAsyncEnumerable<QueryResponse> GetRpcResponsesAsync()
```

That means you'd always need to call one of them, e.g.

```csharp
var lazyQuery = db.RunQueryLazily(...);
foreach (var entity in lazyQuery.GetEntities())
{
   ...
}
```

That's slightly ugly in some ways, but maybe it's appropriate...

Comment (jskeet) [inline apis/Google.Datastore.V1Beta3/Google.Datastore.V1Beta3.Snippets/DatastoreDbSnippets.cs]: As an aside, returning an `IEnumerator<T>` is pretty useless in .NET:

- I can't iterate over it with `foreach`
- I need to remember to dispose of it
- I can't use LINQ with it, or anything else that will use `IEnumerable<T>`

It definitely appeals in terms of expressing the "iterate once", but it's just not idiomatic :(

Comment (SurferJeffAtGoogle) [inline apis/Google.Datastore.V1Beta3/Google.Datastore.V1Beta3.Snippets/DatastoreDbSnippets.cs]: I just came to the same conclusion about IEnumerator<T> after looking up the documentation on foreach.

Comment (SurferJeffAtGoogle) [inline apis/Google.Datastore.V1Beta3/Google.Datastore.V1Beta3.Snippets/DatastoreDbSnippets.cs]: jskeet, I really like this API your proposing, with one minor tweak:
```
DatastoreQueryResults RunQuery(...)
LazyDatastoreQuery RunQueryLazily(...)
```

The `Eagerly` suffix isn't necessary.  When ever I see a function DoFoo(), I assume it does it immediately (eagerly).

Comment (jskeet) [inline apis/Google.Datastore.V1Beta3/Google.Datastore.V1Beta3.Snippets/DatastoreDbSnippets.cs]: Okay. We'd have `RunQueryAsync` as well, returning a `Task<DatastoreQueryResults>`. I'll try to get a new version of this PR done tomorrow - it'll probably just be an extra commit on top of this one.

**Comment by jskeet** [comment]
Okay, it's now *mostly* as we discussed :) Each of `DatastoreDb` and `DatastoreTransaction` has 8 methods:

- `DatastoreQueryResults RunQuery(Query query, ...)`
- `Task<DatastoreQueryResults> RunQueryAsync(Query query, ...)`
- `DatastoreQueryResults RunQuery(GqlQuery query, ...)`
- `Task<DatastoreQueryResults> RunQueryAsync(GqlQuery query, ...)`
- `LazyDatastoreQuery RunQueryLazily(Query query, ...)`
- `AsyncLazyDatastoreQueryResults RunQueryLazilyAsync(Query query, ...)`
- `LazyDatastoreQuery RunQueryLazily(GqlQuery query, ...)`
- `AsyncLazyDatastoreQueryResults RunQueryLazilyAsync(GqlQuery query, ...)`

In theory we *could* have just two methods for lazy evaluation, and determine the async part later... but that turned out to be a little more complex to implement, harder to fake, and no more usable IMO. Feel free to push back on that part...

I've added snippets for the eager execution as well.

**Comment by SurferJeffAtGoogle** [comment]
LGTM.

**Comment by jskeet** [comment]
Merging as at least this is better than what we had. More room for tweaking later, if necessary :)
